go/arrow/decimal256/decimal256.go

```diff
@@ -0,0 +1,569 @@
+	}
+
+	if v < 0 {
+		dec, err := fromPositiveFloat64(-v, prec, scale)
+		if err != nil {
+			return dec, err
+		}
+		return dec.Negate(), nil
+	}
+	return fromPositiveFloat64(v, prec, scale)
+}
+
+func fromPositiveFloat64(v float64, prec, scale int32) (Num, error) {
+	var pscale float64
+	if scale >= -76 && scale <= 76 {
+		pscale = float64PowersOfTen[scale+76]
+	} else {
+		pscale = math.Pow10(int(scale))
+	}
+
+	v *= pscale
+	v = math.RoundToEven(v)
+	maxabs := float64PowersOfTen[prec+76]
+	if v <= -maxabs || v >= maxabs {
+		return Num{}, fmt.Errorf("cannot convert %f to decimal256(precision=%d, scale=%d): overflow",
+			v, prec, scale)
+	}
+
+	var arr [4]float64
+	arr[3] = math.Floor(math.Ldexp(v, -192))
+	v -= math.Ldexp(arr[3], 192)
+	arr[2] = math.Floor(math.Ldexp(v, -128))
+	v -= math.Ldexp(arr[2], 128)
+	arr[1] = math.Floor(math.Ldexp(v, -64))
+	v -= math.Ldexp(arr[1], 64)
+	arr[0] = v
+
+	debug.Assert(arr[3] >= 0, "bad conversion float64 to decimal256")
+	debug.Assert(arr[3] < 1.8446744073709552e+19, "bad conversion float64 to decimal256") // 2**64
+	debug.Assert(arr[2] >= 0, "bad conversion float64 to decimal256")
+	debug.Assert(arr[2] < 1.8446744073709552e+19, "bad conversion float64 to decimal256") // 2**64
+	debug.Assert(arr[1] >= 0, "bad conversion float64 to decimal256")
+	debug.Assert(arr[1] < 1.8446744073709552e+19, "bad conversion float64 to decimal256") // 2**64
+	debug.Assert(arr[0] >= 0, "bad conversion float64 to decimal256")
+	debug.Assert(arr[0] < 1.8446744073709552e+19, "bad conversion float64 to decimal256") // 2**64
+	return Num{[4]uint64{uint64(arr[0]), uint64(arr[1]), uint64(arr[2]), uint64(arr[3])}}, nil
+}
+
+func (n Num) tofloat32Positive(scale int32) float32 {
+	const twoTo64 float32 = 1.8446744e+19
+	if n.arr[3] != 0 || n.arr[2] != 0 {
+		return floatInf
+	}
+	x := float32(n.arr[1]) * twoTo64
+	x += float32(n.arr[0])
+	if scale >= -76 && scale <= 76 {
+		return x * float32PowersOfTen[-scale+76]
+	}
+
+	return x * float32(math.Pow10(-int(scale)))
+}
+
+func (n Num) tofloat64Positive(scale int32) float64 {
+	const (
+		twoTo64  float64 = 1.8446744073709552e+19
+		twoTo128 float64 = 3.402823669209385e+38
+		twoTo192 float64 = 6.277101735386681e+57
+	)
+
+	x := float64(n.arr[3]) * twoTo192
+	x += float64(n.arr[2]) * twoTo128
+	x += float64(n.arr[1]) * twoTo64
+	x += float64(n.arr[0])
+
+	if scale >= -76 && scale <= 76 {
+		return x * float64PowersOfTen[-scale+76]
+	}
+
+	return x * math.Pow10(-int(scale))
+}
+
+func (n Num) ToFloat32(scale int32) float32 {
+	if n.Sign() < 0 {
+		return -n.Negate().tofloat32Positive(scale)
+	}
+	return n.tofloat32Positive(scale)
+}
+
+func (n Num) ToFloat64(scale int32) float64 {
+	if n.Sign() < 0 {
+		return -n.Negate().tofloat64Positive(scale)
+	}
+	return n.tofloat64Positive(scale)
+}
+
+func (n Num) Sign() int {
+	if n == (Num{}) {
+		return 0
+	}
+	return int(1 | (int64(n.arr[3]) >> 63))
+}
+
+func FromBigInt(v *big.Int) (n Num) {
+	bitlen := v.BitLen()
+	if bitlen > 256 {
```

Comment:
Shouldn't you account for the sign bit?
```suggestion
	if bitlen > 255 {
```

Reply:
that's a good point. `big.Int` actually stores the sign as a separate bool with the value, so the `BitLen()` call wouldn't account for the sign bit already, I was assuming it would.